ephios/core/utils.py

```diff
@@ -19,3 +27,26 @@ def password_changed(self, password, user):
             )
             message.attach_alternative(html_content, "text/html")
             message.send()
+
+
+def send_participation_finished(sender, **kwargs):
+    """
+    This method is registered in signals.py as a receiver of ``periodic_signal``.
+    """
+    from ephios.core.signals import participation_finished
+
+    relevant_participations = AbstractParticipation.objects.filter(
+        state=AbstractParticipation.States.CONFIRMED,
+        shift__end_time__lt=timezone.now(),
+        finished=False,
+    ).select_related("shift", "shift__event")
+    for participation in relevant_participations:
+        try:
```

Comment:
this means that it isn't guaranteed that the signal is only called once for the participation, is it?
e.g. the first receiver sends out some notifications, the second receiver raises. The participation will not be marked as finished and the signal will be called again for it the next time

Reply:
That's correct, though I first intended to do it differently. 
Currently, everything is wrapped in a transaction, so that should roll back changes of the first receiver, depending on what it did.
There's definitely a discussion here.

* Single try, no transaction: good for stuff that calls external APIs or something similar that won't be rolled back with a transaction. Though I think these calls should be idempotent anyway.
* Multiple tries, transaction rollback: Somewhat better for stuff like that works with the DB, like models (notifications), although they could do their own handling, and they would probably fail again?!

 I'm not sure, I would be fine with keeping it as is. We could also use the signal's `sender` argument to let receivers decide:
 
 ```python
 try:
    with transaction.atomic():
        participation_finished.send(TRANSACTION, participation=participation)
        participation.finished = True
        participation.save()
except Exception:
    ...
else:
    participation_finished.send(SINGLE_CALL, participation=participation)
 ```
 
 What do you think?

Reply:
I think I'm also fine with it for the moment

Reply:
I will add a notice in the signal's docstring.

Reply:
I did decide to switch to the single call.